File: tmb_syntax/check_simulations.cpp
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_STRING(distr);
  DATA_INTEGER(n);
  Type ans = 0;

  if (distr == "norm") {
    PARAMETER(mu);
    PARAMETER(sd);
    vector<Type> x = rnorm(n, mu, sd);
    ans -= dnorm(x, mu, sd, true).sum();
  }
  else if (distr == "gamma") {
    PARAMETER(shape);
    PARAMETER(scale);
    vector<Type> x = rgamma(n, shape, scale);
    ans -= dgamma(x, shape, scale, true).sum();
  }
  else if (distr == "pois") {
    PARAMETER(lambda);
    vector<Type> x = rpois(n, lambda);
    ans -= dpois(x, lambda, true).sum();
  }
  else if (distr == "compois") {
    PARAMETER(mode);
    PARAMETER(nu);
    vector<Type> x = rcompois(n, mode, nu);
    ans -= dcompois(x, mode, nu, true).sum();
  }
  else if (distr == "compois2") {
    PARAMETER(mean);
    PARAMETER(nu);
    vector<Type> x = rcompois2(n, mean, nu);
    ans -= dcompois2(x, mean, nu, true).sum();
  }
  else if (distr == "tweedie") {
    PARAMETER(mu);
    PARAMETER(phi);
    PARAMETER(p);
    vector<Type> x = rtweedie(n, mu, phi, p);
    ans -= dtweedie(x, mu, phi, p, true).sum();
  }
  else if (distr == "nbinom") {
    PARAMETER(size);
    PARAMETER(prob);
    vector<Type> x = rnbinom(n, size, prob);
    ans -= dnbinom(x, size, prob, true).sum();
  }
  else if (distr == "nbinom2") {
    PARAMETER(mu);
    PARAMETER(var);
    vector<Type> x = rnbinom2(n, mu, var);
    ans -= dnbinom2(x, mu, var, true).sum();
  }
  else if (distr == "exp") {
    PARAMETER(rate);
    vector<Type> x = rexp(n, rate);
    ans -= dexp(x, rate, true).sum();
  }
  else if (distr == "beta") {
    PARAMETER(shape1);
    PARAMETER(shape2);
    vector<Type> x = rbeta(n, shape1, shape2);
    ans -= dbeta(x, shape1, shape2, true).sum();
  }
  else if (distr == "f") {
    PARAMETER(df1);
    PARAMETER(df2);
    vector<Type> x = rf(n, df1, df2);
    ans -= df(x, df1, df2, true).sum();
  }
  else if (distr == "logis") {
    PARAMETER(location);
    PARAMETER(scale);
    vector<Type> x = rlogis(n, location, scale);
    ans -= dlogis(x, location, scale, true).sum();
  }
  else if (distr == "t") {
    PARAMETER(df);
    vector<Type> x = rt(n, df);
    ans -= dt(x, df, true).sum();
  }
  else if (distr == "weibull") {
    PARAMETER(shape);
    PARAMETER(scale);
    vector<Type> x = rweibull(n, shape, scale);
    ans -= dweibull(x, shape, scale, true).sum();
  }
  else if (distr == "AR1") {
    PARAMETER(phi);
    vector<Type> x(n);
    density::AR1(phi).simulate(x);
    ans += density::AR1(phi)(x);
  }
  else if (distr == "ARk") {
    PARAMETER_VECTOR(phi);
    vector<Type> x(n);
    density::ARk(phi).simulate(x);
    ans += density::ARk(phi)(x);
  }
  else if (distr == "MVNORM") {
    PARAMETER(phi);
    matrix<Type> Sigma(5,5);
    for(int i=0; i<Sigma.rows(); i++)
      for(int j=0; j<Sigma.rows(); j++)
        Sigma(i,j) = exp( -phi * abs(i - j) );
    density::MVNORM_t<Type> nldens = density::MVNORM(Sigma);
    for(int i = 0; i<n; i++) {
      vector<Type> x = nldens.simulate();
      ans += nldens(x);
    }
  }
  else if (distr == "SEPARABLE") {
    PARAMETER(phi1);
    PARAMETER_VECTOR(phi2);
    array<Type> x(100, 200);
    SEPARABLE( density::ARk(phi2), density::AR1(phi1) ).simulate(x);
    ans += SEPARABLE( density::ARk(phi2), density::AR1(phi1) )(x);
  }
  else if (distr == "GMRF") {
    PARAMETER(delta);
    matrix<Type> Q0(5, 5);
    Q0 <<
      1,-1, 0, 0, 0,
     -1, 2,-1, 0, 0,
      0,-1, 2,-1, 0,
      0, 0,-1, 2,-1,
      0, 0, 0,-1, 1;
    Q0.diagonal().array() += delta;
    Eigen::SparseMatrix<Type> Q = asSparseMatrix(Q0);
    vector<Type> x(5);
    for(int i = 0; i<n; i++) {
      density::GMRF(Q).simulate(x);
      ans += density::GMRF(Q)(x);
    }
  }
  else if (distr == "SEPARABLE_NESTED") {
    PARAMETER(phi1);
    PARAMETER(phi2);
    PARAMETER(delta);
    matrix<Type> Q0(5, 5);
    Q0 <<
      1,-1, 0, 0, 0,
     -1, 2,-1, 0, 0,
      0,-1, 2,-1, 0,
      0, 0,-1, 2,-1,
      0, 0, 0,-1, 1;
    Q0.diagonal().array() += delta;
    Eigen::SparseMatrix<Type> Q = asSparseMatrix(Q0);
    array<Type> x(5, 6, 7);
    for(int i = 0; i<n; i++) {
      SEPARABLE(density::AR1(phi2),
                SEPARABLE(density::AR1(phi1),
                          density::GMRF(Q) ) ).simulate(x);
      ans += SEPARABLE(density::AR1(phi2),
                       SEPARABLE(density::AR1(phi1),
                                 density::GMRF(Q) ) )(x);
    }
  }
  else error( ("Invalid distribution '" + distr + "'").c_str() );
  return ans;
}

File: tmb_syntax/check_simulations.R

require(TMB)
compile('check_simulations.cpp')
dyn.load(dynlib('check_simulations'))
set.seed(1)

.results <- list()
checkSimulation <- function(obj) {
    g <- obj$gr(obj$par) ## Gradient at true parameter
    H <- obj$he(obj$par) ## Hessian at true parameter
    stopifnot(all( eigen(H)$value > 0 ))
    ## g ~ N(0, H)
    chi.sq <- g %*% solve(H) %*% t(g)
    p.value <- 1 - pchisq(chi.sq, df=length(obj$par))
    ans <- data.frame(distribution=obj$env$data$distr,
                      p.value=p.value, PASSED = p.value>.05 )
    .GlobalEnv$.results <- rbind(.GlobalEnv$.results, ans)
    print(ans)
    stopifnot(ans$PASSED)
    invisible(ans)
}

n <- 100000

######################################################################

data <- list(distr="norm", n=n)
parameters <- list(mu=0, sd=1)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="gamma", n=n)
parameters <- list(shape=3, scale=2)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="pois", n=n)
parameters <- list(lambda=5)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="compois", n=n)
parameters <- list(mode=3, nu=0.5)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="tweedie", n=n)
parameters <- list(mu=10, phi=10, p=1.5)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="compois2", n=n)
parameters <- list(mean=3, nu=0.1)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="nbinom", n=n)
parameters <- list(size=5, prob=.1)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="nbinom2", n=n)
parameters <- list(mu=5, var=10)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="exp", n=n)
parameters <- list(rate=1)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="beta", n=n)
parameters <- list(shape1=0.5, shape2=0.5)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="f", n=n)
parameters <- list(df1=5, df2=5)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="logis", n=n)
parameters <- list(location=0, scale=1)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="t", n=n)
parameters <- list(df=5)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="weibull", n=n)
parameters <- list(shape=1, scale=1)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################
data <- list(distr="AR1", n=n)
parameters <- list(phi = .2)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="ARk", n=n)
parameters <- list(phi = c(.9, .05))
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

data <- list(distr="MVNORM", n=1000)
parameters <- list(phi = c(.9))
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################
data <- list(distr="SEPARABLE", n=n)
parameters <- list(phi1 = .2, phi2 = c(.9, .05))
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################
data <- list(distr="GMRF", n=1000)
parameters <- list(delta = .1)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################
data <- list(distr="SEPARABLE_NESTED", n=1000)
parameters <- list(phi1 = .2, phi2 = .9, delta = .1)
obj <- MakeADFun(data, parameters)
checkSimulation(obj)

######################################################################

cat("\n"); print(.results)
